From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: zfcp: return early from slave_destroy if slave_alloc returned early
Patch-mainline: yes
References: bnc#732371,LTC#76754

Symptom:     Kernel page faults on accessing uninitialized fields of
             struct zfcp_scsi_dev in zfcp_erp_lun_shutdown_wait.
Problem:     zfcp_scsi_slave_destroy erroneously always tried to finish its
             task even if the corresponding previous zfcp_scsi_slave_alloc
             returned early.
Solution:    Take the port field of struct zfcp_scsi_dev to determine
             if slave_alloc returned early.

Acked-by: John Jolly <jjolly@suse.de>
---
 drivers/s390/scsi/zfcp_scsi.c |    4 ++++
 1 file changed, 4 insertions(+)

--- a/drivers/s390/scsi/zfcp_scsi.c
+++ b/drivers/s390/scsi/zfcp_scsi.c
@@ -54,6 +54,10 @@ static void zfcp_scsi_slave_destroy(stru
 {
 	struct zfcp_scsi_dev *zfcp_sdev = sdev_to_zfcp(sdev);
 
+	/* if previous slave_alloc returned early, there is nothing to do */
+	if (!zfcp_sdev->port)
+		return;
+
 	zfcp_erp_lun_shutdown_wait(sdev, "scssd_1");
 	put_device(&zfcp_sdev->port->dev);
 }
